Final attempt

PiKeyAr opened the pull request:
List of changes:
- Updated SDL2 to 2.0.12

- Reworked keyboard-related functions to update vanilla SADX key array

- Reworked keyboard-related functions to support keyboard mappings read from `config.ini`

- Added a global toggle to disable mouse input

- Added "center camera" (E key in vanilla SADX) for all controllers ("left stick" button) and the keyboard

- Added "half press" functionality like on the keyboard for controllers ("right stick" button)

- Support for soft reset by pressing ALT+F2 or ABXY+Start

- Updated `configschema.xml` with capitalized `True` and `False` for better compatibility with the Mod Manager's config editor

- Implemented camera scrolling with the mouse wheel

michael-fadely commented on sadx-input-mod/SADXKeyboard.h:
```diff
@@ -0,0 +1,123 @@
+	Uint32 SADX2004Code;
+};
+
+KeyData SADX2004Keys[] = {
```
This should be defined in a cpp file and marked as `extern` (e.g. `extern KeyData SADX2004Keys[];`)

michael-fadely commented on sadx-input-mod/KeyboardMouse.cpp:
```diff
@@ -2,22 +2,50 @@
+
+void KeyboardMouse::update_sadx_key(Uint32 key, bool down)
+{
+	for (int i = 0; i < LengthOfArray(SADX2004Keys); i++)
+	{
+		if (key == SADX2004Keys[i].WindowsCode)
+		{
+			if (KeyboardKeys[SADX2004Keys[i].SADX2004Code].old != down) KeyboardKeys[SADX2004Keys[i].SADX2004Code].pressed = down;
+			KeyboardKeys[SADX2004Keys[i].SADX2004Code].old = KeyboardKeys[SADX2004Keys[i].SADX2004Code].held;
+			KeyboardKeys[SADX2004Keys[i].SADX2004Code].held = down;
+			return;
+		}
+	}
```
Given the way it is used, I would suggest that `SADX2004Keys` be changed to `std::map`. Its definition would look something like:
```cpp
std::map<int, int> SADX2004Keys
```

And this entire for loop can be completely removed. e.g.:
```cpp
auto it = SADX2004Keys.find(key);

// no equivalent windows -> sadx mapping available
if (it == SADX2004Keys.end())
{
    return;
}

auto& keyboard_key = KeyboardKeys[it->second];

if (keyboard_key.old != down) keyboard_key.pressed = down;
keyboard_key.old = keyboard_key.held;
keyboard_key.held = down;
```

michael-fadely commented on sadx-input-mod/input.cpp:
```diff
@@ -19,7 +19,9 @@ namespace input
 	bool controller_enabled[GAMEPAD_COUNT];
 	bool debug = false;
-
+	bool disablemouse = true;
```
Preferred naming style would make this `disable_mouse`

michael-fadely commented on sadx-input-mod/input.cpp:
```diff
@@ -19,7 +19,7 @@ namespace input
 	bool debug = false;
-	bool disablemouse = true;
+	bool disable_mouse = true;
 	bool e_held;
```
This should be default-initialized. `false` would make sense to me.

michael-fadely commented on sadx-input-mod/mod.cpp:
```diff
@@ -15,6 +15,8 @@
 #include "input.h"
 #include "rumble.h"
 
+DataPointer(char, SoftResetByte, 0x3B0EAA0);
```
Not a deal-breaker, but it'd be nice to avoid duplicate code. Maybe a header for variable definitions.

PiKeyAr commented:
I've added compatibility with SADXPC demo player (disabled in vanilla but restored in DC Conversion), tried to do it as cleanly as I could but this probably needs additional review.

PiKeyAr commented:
Updated vanilla SADX keyboard array handling to work much better.

michael-fadely commented:
I may be slow to review, but I will do it for sure. I hope to get it done this week. Continue making changes as you are.

michael-fadely commented on sadx-input-mod/DreamPad.cpp:
```diff
@@ -138,7 +138,7 @@ void DreamPad::poll()
 
 		normalized_l_ = convert_axes(reinterpret_cast<NJS_POINT2I*>(&dc_pad.LeftStickX), axis, settings.deadzone_l,
-		                             settings.radial_l);
+						settings.radial_l);
```
Spaces are preferred for alignment, with tabs used for indentation level.

michael-fadely commented on sadx-input-mod/DreamPad.cpp:
```diff
@@ -155,7 +155,7 @@ void DreamPad::poll()
 
 		normalized_r_ = convert_axes(reinterpret_cast<NJS_POINT2I*>(&dc_pad.RightStickX), axis, settings.deadzone_r,
-		                             settings.radial_r);
+						settings.radial_r);
```
ditto

michael-fadely commented on sadx-input-mod/DreamPad.cpp:
```diff
@@ -248,6 +250,30 @@ void DreamPad::poll()
+		dc_pad.RightStickY = 0;
+		dc_pad.LTriggerPressure = 0;
+		dc_pad.RTriggerPressure = 0;
+		if (buttons & Buttons_Start) buttons = Buttons_Start;
+		else buttons = 0;
```
Inline if preferred, e.g. `buttons = (buttons & Buttons_Start) ? Buttons_Start : 0;`

michael-fadely commented on sadx-input-mod/mod.cpp:
```diff
@@ -66,7 +135,7 @@ extern "C"
 
 			MessageBoxA(nullptr, "Error loading SDL. See debug message for details.",
-			            "SDL Load Error", MB_OK | MB_ICONERROR);
+				"SDL Load Error", MB_OK | MB_ICONERROR);
```
Spaces for alignment preferred.

michael-fadely commented on sadx-input-mod/mod.cpp:
```diff
@@ -76,12 +145,13 @@ extern "C"
 			PrintDebug("[Input] Unable to initialize SDL. Error code: %i\n", init);
 			MessageBoxA(nullptr, "Error initializing SDL. See debug message for details.",
-			            "SDL Init Error", MB_OK | MB_ICONERROR);
+				"SDL Init Error", MB_OK | MB_ICONERROR);
```
ditto

michael-fadely commented on sadx-input-mod/mod.cpp:
```diff
@@ -163,12 +258,12 @@ extern "C"
 			{
 				PrintDebug("[Input] Deadzones for P%d (L/R/T): %05d / %05d / %05d\n", (i + 1),
-				           settings.deadzone_l, settings.deadzone_r, settings.trigger_threshold);
+					settings.deadzone_l, settings.deadzone_r, settings.trigger_threshold);
```
Spaces for alignment

michael-fadely commented on sadx-input-mod/mod.cpp:
```diff
@@ -163,12 +258,12 @@ extern "C"
 			settings.rumble_min_time = static_cast<ushort>(config.getInt(section, "RumbleMinTime", 0));
 
-			settings.allow_keyboard = config.getBool(section, "AllowKeyboard", !i);
+			if (i == KeyboardPlayer) settings.allow_keyboard = true; else settings.allow_keyboard = false;
```
`settings.allow_keyboard = (i == KeyboardPlayer);`

PiKeyAr commented:
I replaced all tabs with spaces in parts you mentioned, still not sure if it looks correct though.
Also changed ifs to inline stuff where you indicated.

michael-fadely commented:
They should use tabs for indentation level, then spaces to align with the first opening parenthesis, like so:

![image](https://user-images.githubusercontent.com/920569/81524703-69f9ea00-930f-11ea-9297-0152af2e102c.png)

Lines indicate tabs, dots indicate spaces.

PiKeyAr commented:
OK, done

michael-fadely commented:
Close, but you've got one indentation level (tab) too many on all of them. They should have the same number of tabs as the previous line.
